### PR TITLE
Allow DesiredCapacity to be configured via parameter

### PR DESCRIPTION
**Pull request by deppy**
Allow the DesiredCapacity to be configured via parameter to prevent always scaling down to the MinSize during stack updates. (Issue #260)

### CODE REVIEW

**Comment by toolmantim** [comment]
@deppy I just found a problem/side-effect of this new addition. If you set `MinSize` to a non-zero value (say, 1), but don't forget to set `DesiredCapacity` then you get a stack creation failure after it goes and creates a bunch of resources. And because it's not a validation error, you have to go through the whole stack creation process again.

I've poured over the CloudFormation documentation to try to do:

```
DesiredCapacity: $(Maximum of (DesiredCapacity || MinSize)
```

or:

```
DesiredCapacity: $(DesiredCapacity < MinSize ? MinSize : DesiredCapacity)
```

or just to add support for validating that `MinSize >= DesiredCapacity`

… but I don't think any of these are possible.

Given we can't validate it, this makes the stack setup process much less easy-to-use :(

Is there another solution we could switch to, to have the same affect? I think this new parameter could add a lot of confusion.

**Comment by deppy** [comment]
Fair Point.  DesiredCapacity is an optional field.  I believe the behavior would be that the initial value would be set to MinSize, and then on upgrade it shouldn't modify it (ie. solving the problem that I was running into where an update would downscale to MinValue).  I've put in Support Request to amazon to see what the default value would be (I have no time to test that at the moment to find out the answer).

If it does indeed behave that way, I'd be up for just removing DesiredCapacity from the template since it would revert back to the previous behavior of initially being the MinSize, while fixing the downscale issue we were facing.  

Will let you know when I get feedback from AWS.

**Comment by toolmantim** [comment]
Oh yeah, I hadn't considered removing it altogether! That could work. Thanks for reaching out to AWS to figure it out.

**Comment by SomeoneWeird** [comment]
@deppy Did you hear back from AWS?

**Comment by lox** [comment]
So shall we just remove it? @SomeoneWeird points out it's actually a breaking change as it stands.

**Comment by deppy** [comment]
I had a few messages back and forth with AWS.  They wouldn't directly answer my questions when I gave them multiple scenarios but this is what I've gathered so far....

If you don't include DesiredCapacity, on initial stack creation it will use the MinSize value.  On stack update it will leave it as is the current value for DesiredCapacity unless the stack update requires a replacement of the ASG, at which point it will revert back to the MinSize (which was the case that was breaking for me, because we often change the MaxSize).

As for the other part of verifying that `MinSize <= DesiredCapcity >= MaxSize`, they basically said there is no way to do that.  I've given them a feature request for this, that would require that they allow you to use a parameter as a value for another parameters MinValue and MaxValue.

Basically there's no way to validate that the value someone entered was in the correct range and the only way to not revert back to the MinSize for DesiredCapacity is do declare it when replacing the ASG in the stack update.  So feel free to revert this and I can just keep handling it the way I have been, which is manually updating the DesiredCapacity after a stack update.  Less than ideal, but it will at least prevent others from putting invalid values and breaking their stack.

**Comment by SomeoneWeird** [comment]
```
If you don't include DesiredCapacity, on initial stack creation it will use the MinSize value.
```

This doesn't seem to be the case, currently. The stack exits with an error if `DesiredCapacity` is not defined.

**Comment by deppy** [comment]
@SomeoneWeird By "If you don't include DesiredCapacity", I mean that if you completely pull it out of the template.  The way it exists now is because it doesn't have a default because it must it must bet MinSize <= DesiredCapacity >= MaxSize.

**Comment by toolmantim** [comment]
What if we set `DesiredCapacity` to `MaxSize`? That way when you update or create the stack it scales out, but then scales in afterwards if there's no jobs. And stack updates will be much smoother than reverting to the old behavior.

**Comment by deppy** [comment]
@toolmantim That works for me :-)

**Comment by lox** [comment]
https://github.com/buildkite/elastic-ci-stack-for-aws/pull/270 changes things so that if you leave it at zero it will be the empty value. I think that captures the above?
